File: scripts/restore-kv.mjs
#!/usr/bin/env node
/**
 * Restore KV Namespace from JSON Backup
 * 
 * Restores all keys and values from a backup JSON file to TASKS_KV.
 * 
 * Usage:
 *   node scripts/restore-kv.mjs <backup-file>
 * 
 * Example:
 *   node scripts/restore-kv.mjs backups/tasks-kv-backup-2025-10-14T18-30-00.json
 */

import { execSync } from 'child_process';
import fs from 'fs';

const KV_NAMESPACE = 'TASKS_KV';

// Get backup file from command line argument
const backupFile = process.argv[2];

if (!backupFile) {
  console.error('❌ Error: Please provide a backup file path');
  console.log('\nUsage:');
  console.log('  node scripts/restore-kv.mjs <backup-file>');
  console.log('\nExample:');
  console.log('  node scripts/restore-kv.mjs backups/tasks-kv-backup-2025-10-14T18-30-00.json');
  process.exit(1);
}

if (!fs.existsSync(backupFile)) {
  console.error(`❌ Error: Backup file not found: ${backupFile}`);
  process.exit(1);
}

console.log('🔄 Starting KV restore...\n');

try {
  // Load backup file
  console.log(`📂 Loading backup from: ${backupFile}`);
  const backupContent = fs.readFileSync(backupFile, 'utf-8');
  const backup = JSON.parse(backupContent);
  
  console.log(`✅ Backup loaded successfully`);
  console.log(`   Timestamp: ${backup.timestamp}`);
  console.log(`   Namespace: ${backup.namespace}`);
  console.log(`   Key count: ${backup.keyCount}\n`);
  
  if (!backup.data || Object.keys(backup.data).length === 0) {
    console.log('⚠️  No data found in backup file');
    process.exit(0);
  }
  
  // Restore each key-value pair
  console.log('📦 Restoring key-value pairs...');
  const entries = Object.entries(backup.data);
  let count = 0;
  let failed = 0;
  
  for (const [key, value] of entries) {
    try {
      // Convert value to string (JSON if object, otherwise as-is)
      const valueStr = typeof value === 'string' 
        ? value 
        : JSON.stringify(value);
      
      // Escape quotes for shell command
      const escapedValue = valueStr.replace(/"/g, '\\"');
      
      const putCommand = `npx wrangler kv key put "${key}" "${escapedValue}" --binding=${KV_NAMESPACE}`;
      execSync(putCommand, { 
        cwd: 'workers/task-api',
        stdio: 'pipe' // Suppress output
      });
      
      count++;
      process.stdout.write(`\r   Progress: ${count}/${entries.length} keys restored`);
    } catch (error) {
      failed++;
      console.error(`\n❌ Failed to restore key: ${key}`, error.message);
    }
  }
  
  console.log('\n');
  console.log(`✅ Restore completed!`);
  console.log(`   Successfully restored: ${count} keys`);
  if (failed > 0) {
    console.log(`   Failed: ${failed} keys`);
  }
  
  console.log('\n✨ KV namespace restored from backup!');
  
} catch (error) {
  console.error('\n❌ Restore failed:', error.message);
  process.exit(1);
}

File: scripts/restore-kv.py
#!/usr/bin/env python3
"""KV Restore Script"""
import requests, json, os, sys
from pathlib import Path

def load_config():
    env_path = Path(__file__).parent.parent / '.env'
    config = {'CLOUDFLARE_API_TOKEN': os.environ.get('CLOUDFLARE_API_TOKEN'),
              'CLOUDFLARE_ACCOUNT_ID': os.environ.get('CLOUDFLARE_ACCOUNT_ID'),
              'CLOUDFLARE_NAMESPACE_ID': os.environ.get('CLOUDFLARE_NAMESPACE_ID')}
    if not all(config.values()) and env_path.exists():
        with open(env_path) as f:
            for line in f:
                if '=' in line and not line.strip().startswith('#'):
                    k, v = line.split('=', 1)
                    k, v = k.strip(), v.strip().strip('"').strip("'")
                    if k in config and not config[k]: config[k] = v
    return config

def main(backup_file):
    print(f"Restoring from: {backup_file}")
    cfg = load_config()
    if not all([cfg['CLOUDFLARE_API_TOKEN'], cfg['CLOUDFLARE_ACCOUNT_ID'], cfg['CLOUDFLARE_NAMESPACE_ID']]):
        print("Error: Missing credentials"); return 1
    
    with open(backup_file) as f: data = json.load(f)
    keys = data['keys']
    print(f"Restoring {len(keys)} keys...")
    
    headers = {'Authorization': f'Bearer {cfg['CLOUDFLARE_API_TOKEN']}'}
    for i, (key, value) in enumerate(keys.items(), 1):
        url = f"https://api.cloudflare.com/client/v4/accounts/{cfg['CLOUDFLARE_ACCOUNT_ID']}/storage/kv/namespaces/{cfg['CLOUDFLARE_NAMESPACE_ID']}/values/{key}"
        val_str = json.dumps(value) if isinstance(value, (dict, list)) else str(value)
        r = requests.put(url, headers=headers, data=val_str.encode('utf-8'))
        print(f"Progress: {i}/{len(keys)}", end='\r')
    
    print(f"\nRestore complete: {len(keys)} keys")
    return 0

if __name__ == '__main__':
    if len(sys.argv) < 2: print("Usage: python restore-kv.py <backup-file>"); sys.exit(1)
    sys.exit(main(sys.argv[1]))
